### PR TITLE
improvement(_pytest/assertion/util.py): improve visual perception of comparison report on dataclasses / attrs

### LINKED ISSUE #7348: Improve recursive diff output for comparison asserts on dataclasses/attrs
pytest 6 will have a new feature (PR #6835) for displaying nicer diff output on large dataclasses. Often dataclasses have other dataclasses or other types nested inside them, and the assertion fails because one or more deeply nested fields differ. Before the diff was just on the top-level `repr` of the objects, which was very hard to decipher for larger dataclasses. After the PR, the output looks like this (see https://github.com/pytest-dev/pytest/pull/6835#issuecomment-637545552):

```
    def test_cmp():
>       assert obj == obj.replace(assentior=obj.assentior.replace(culpa=7, aliquip=Aliquip.Aliquip3))
E       AssertionError: assert Dignissim(assentior=Assentior(aliquip=<Aliquip.Aliquip1: 1>, culpa=6, fugiat=Fugiat(eiusmod='aaaaaaaaaaaaaaaa', tempor='bbbbbb', incididunt='CCCCCCCCCCCCCCCCCCC', labore='dddddddddd', dolore='eeeeeeeeee', magna='fffffffffff', aliqua='gggggggggggggg', veniam=None, nostrud=None, exercitation=None, ullamco=None, laboris=None, commodo=None, consequat=None, aute=None), invenire=Invenire(irure=53, reprehenderit=153, voluptate=242, velit=100, esse=5035, cillum=53, eepcillum=422), deserunt=True, lorem=Lorem(ipsum=b'', dolor=['xx'], sit=['yy'], amet=['zz'], consectetur=b'sdfsd', adipiscing=b'adfsdf'), mollit=4294967294, laborums=[Laborum(type=23, urangulal=b'abc', ipsumal=None, occaecat=b'sdf', cupidatat=13, proident=None)], tantas='sdfsdlxcv49249sdfs90sdf==', nominati=-1, fabulas=False, tritani=None), new_cupidatat=13, laoreet=1, rationibus=False) == Dignissim(assentior=Assentior(aliquip=<Aliquip.Aliquip3: 3>, culpa=7, fugiat=Fugiat(eiusmod='aaaaaaaaaaaaaaaa', tempor='bbbbbb', incididunt='CCCCCCCCCCCCCCCCCCC', labore='dddddddddd', dolore='eeeeeeeeee', magna='fffffffffff', aliqua='gggggggggggggg', veniam=None, nostrud=None, exercitation=None, ullamco=None, laboris=None, commodo=None, consequat=None, aute=None), invenire=Invenire(irure=53, reprehenderit=153, voluptate=242, velit=100, esse=5035, cillum=53, eepcillum=422), deserunt=True, lorem=Lorem(ipsum=b'', dolor=['xx'], sit=['yy'], amet=['zz'], consectetur=b'sdfsd', adipiscing=b'adfsdf'), mollit=4294967294, laborums=[Laborum(type=23, urangulal=b'abc', ipsumal=None, occaecat=b'sdf', cupidatat=13, proident=None)], tantas='sdfsdlxcv49249sdfs90sdf==', nominati=-1, fabulas=False, tritani=None), new_cupidatat=13, laoreet=1, rationibus=False)
E         Matching attributes:
E         ['new_cupidatat', 'laoreet', 'rationibus']
E         Differing attributes:
E         assentior: Assentior(aliquip=<Aliquip.Aliquip1: 1>, culpa=6, fugiat=Fugiat(eiusmod='aaaaaaaaaaaaaaaa', tempor='bbbbbb', incididunt='CCCCCCCCCCCCCCCCCCC', labore='dddddddddd', dolore='eeeeeeeeee', magna='fffffffffff', aliqua='gggggggggggggg', veniam=None, nostrud=None, exercitation=None, ullamco=None, laboris=None, commodo=None, consequat=None, aute=None), invenire=Invenire(irure=53, reprehenderit=153, voluptate=242, velit=100, esse=5035, cillum=53, eepcillum=422), deserunt=True, lorem=Lorem(ipsum=b'', dolor=['xx'], sit=['yy'], amet=['zz'], consectetur=b'sdfsd', adipiscing=b'adfsdf'), mollit=4294967294, laborums=[Laborum(type=23, urangulal=b'abc', ipsumal=None, occaecat=b'sdf', cupidatat=13, proident=None)], tantas='sdfsdlxcv49249sdfs90sdf==', nominati=-1, fabulas=False, tritani=None) != Assentior(aliquip=<Aliquip.Aliquip3: 3>, culpa=7, fugiat=Fugiat(eiusmod='aaaaaaaaaaaaaaaa', tempor='bbbbbb', incididunt='CCCCCCCCCCCCCCCCCCC', labore='dddddddddd', dolore='eeeeeeeeee', magna='fffffffffff', aliqua='gggggggggggggg', veniam=None, nostrud=None, exercitation=None, ullamco=None, laboris=None, commodo=None, consequat=None, aute=None), invenire=Invenire(irure=53, reprehenderit=153, voluptate=242, velit=100, esse=5035, cillum=53, eepcillum=422), deserunt=True, lorem=Lorem(ipsum=b'', dolor=['xx'], sit=['yy'], amet=['zz'], consectetur=b'sdfsd', adipiscing=b'adfsdf'), mollit=4294967294, laborums=[Laborum(type=23, urangulal=b'abc', ipsumal=None, occaecat=b'sdf', cupidatat=13, proident=None)], tantas='sdfsdlxcv49249sdfs90sdf==', nominati=-1, fabulas=False, tritani=None)
E         
E         Drill down into differing attribute assentior
E         Matching attributes:
E         ['fugiat',
E          'invenire',
E          'deserunt',
E          'lorem',
E          'mollit',
E          'laborums',
E          'tantas',
E          'nominati',
E          'fabulas',
E          'tritani']
E         Differing attributes:
E         aliquip: <Aliquip.Aliquip1: 1> != <Aliquip.Aliquip3: 3>
E         
E         Drill down into differing attribute aliquip
E         +<Aliquip.Aliquip1: 1>
E         -<Aliquip.Aliquip3: 3>
E         culpa: 6 != 7
E         
E         Drill down into differing attribute culpa
E         +6
E         -7

x.py:169: AssertionError
```

This is a great improvement, but the drill down structure is a bit hard to see. In this case particularly, there are two differing nested attributes, `aliquip` and `culpa`, but the `culpa` one starts just after the output of the `aliquip` without any spacing, which is confusing. Also, it is not visible at which nesting level the differing attribute is.

### PR DESCRIPTION
**Pull request by Vlad-Radz**
improvement(_pytest/assertion/util.py): improve visual perception of comparison report on dataclasses / attrs
chore(test_compare_recursive_dataclasses.py): improve the test case
style(_pytest/assertion/truncate.py): replace some string formatting expressions by f-string

``closes #7348``

### CODE REVIEW

**Comment by Vlad-Radz** [comment]
I am going to fix the tests.

File: src/_pytest/assertion/util.py [hunk 1]
```diff
@@ -96,6 +100,23 @@ def _format_lines(lines: Sequence[str]) -> List[str]:
     return result
 
 
+def _get_number_of_calls(func_name: str) -> int:
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:103]: While this solution is elegant in a way, I don't think using `inspect.stack()` is a good idea. It is a sort-of meta/introspection facility with it its own rules that I think is better to leave for special cases. In this case it should be possible to implement entirely "within" the language.

File: src/_pytest/assertion/util.py [hunk 2]
```diff
@@ -96,6 +100,23 @@ def _format_lines(lines: Sequence[str]) -> List[str]:
...
+    """
+    Get the number of calls for the specified function.
+    To get this, the stack is going to be inspected.
+
+    :param str func_name: the name of the function
+    :return: number of calls from stack
+    :rtype: int
+    :raises ValueError: if the func_name was not found in the stack.s
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:111]: It is more natural in this case to return 0 than to raise.

File: src/_pytest/assertion/util.py [hunk 3]
```diff
@@ -96,6 +100,23 @@ def _format_lines(lines: Sequence[str]) -> List[str]:
...
+    """
+    functions = [frame.function for frame in inspect.stack()]
+    func_calls = Counter(functions).get(func_name, 0)  # int
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:114]: Generally it is possible for multiple different functions in the stack to have the same name, in which case this will get confused.

File: src/_pytest/assertion/util.py [hunk 4]
```diff
@@ -169,23 +190,56 @@ def assertrepr_compare(config, op: str, left: Any, right: Any) -> Optional[List[
 
 
 def _compare_eq_any(left: Any, right: Any, verbose: int = 0) -> List[str]:
-    explanation = []  # type: List[str]
+    """
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:193]: `_compare_eq_any` is an internal function which the reader is kind of expected to understand what it does on their own. I think a one sentence docstring

> Generic comparison function, dispatches to other comparison functions based on the inputs' type.

would be nice, however the rest is mostly redundant with the function signature, so I'd leave it out.

File: src/_pytest/assertion/util.py [hunk 5]
```diff
@@ -169,23 +190,56 @@ def assertrepr_compare(config, op: str, left: Any, right: Any) -> Optional[List[
...
+    :type right: Any
+    :param verbose
+    :type verbose: int
+    .. note:: indendation used for better visual perception of the comparison report.
+    :return: explanation of an AssertionError
+    :rtype: List[str]
+    """
+    explanation = []
+
+    my_name = cast(types.FrameType, inspect.currentframe()).f_code.co_name
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:209]: Instead of using a cast, prefer to assign `inspect.currentframe` to a variable, e.g. `my_frame`, and then `assert my_frame is not None`.

File: src/_pytest/assertion/util.py [hunk 6]
```diff
@@ -169,23 +190,56 @@ def assertrepr_compare(config, op: str, left: Any, right: Any) -> Optional[List[
...
+    num_calls = _get_number_of_calls(func_name=my_name)  # type: int
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:210]: The type comment here and and in the next line shouldn't be needed. Mypy can infer that the type of `num_calls` is `int` from the return value annotation of `_get_number_of_calls`.

File: src/_pytest/assertion/util.py [hunk 7]
```diff
@@ -169,23 +190,56 @@ def assertrepr_compare(config, op: str, left: Any, right: Any) -> Optional[List[
...
+    indentation = " " * 4 * (num_calls - 1)  # type: str
+
     if istext(left) and istext(right):
-        explanation = _diff_text(left, right, verbose)
+        explanation = _diff_text(left=left, right=right, verbose=verbose)
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:214]: Let's keep the arguments positional, here and below:

1. `left=left` is a bit redundant and longer, the positionals are a bit more readable here.
2. It is harder to review a diff when style changes are included, so it's better to submit them in a separate PR.

File: src/_pytest/assertion/util.py [hunk 8]
```diff
@@ -436,16 +506,21 @@ def _compare_eq_cls(
     if same and verbose < 2:
         explanation.append("Omitting %s identical items, use -vv to show" % len(same))
     elif same:
-        explanation += ["Matching attributes:"]
-        explanation += pprint.pformat(same).splitlines()
+        explanation += ["{}Matching attributes:".format(indentation)]
+        mystr = [indentation + pprint.pformat(same).splitlines()[0]]
```
Comment (bluetech) [inline src/_pytest/assertion/util.py:510]: The code changed here -- instead of including all of the lines of `pprint.pformat(same).splitlines()`, it now only includes the first.

Also, it would be better to avoid assigning to `mystr` here, and include it in the line below directly, because `mystr` is not used again.

File: testing/test_assertion.py [hunk 9]
```diff
@@ -797,8 +797,8 @@ def test_recursive_dataclasses(self, testdir):
                 "*Differing attributes:*",
                 "*field_b: ComplexDataObject2(*SimpleDataObject(field_a=2, field_b='c')) != ComplexDataObject2(*SimpleDataObject(field_a=3, field_b='c'))*",  # noqa
                 "*Drill down into differing attribute field_b:*",
-                "*Omitting 1 identical items, use -vv to show*",
                 "*Differing attributes:*",
+                "*field_a: ComplexDataObject*",
```
Comment (bluetech) [inline testing/test_assertion.py:801]: Because the indentation is an important part of the output, the tests should include them in the test, i.e. drop the `*` wildcards at the beginnings. Otherwise the tests accept any indentation, and it is not verified to be correct.

File: testing/test_assertion.py [hunk 10]
```diff
@@ -813,15 +813,14 @@ def test_recursive_dataclasses_verbose(self, testdir):
                 "*Matching attributes:*",
                 "*['field_a']*",
                 "*Differing attributes:*",
-                "*field_b: ComplexDataObject2(*SimpleDataObject(field_a=2, field_b='c')) != ComplexDataObject2(*SimpleDataObject(field_a=3, field_b='c'))*",  # noqa
-                "*Matching attributes:*",
-                "*['field_a']*",
-                "*Differing attributes:*",
-                "*field_b: SimpleDataObject(field_a=2, field_b='c') != SimpleDataObject(field_a=3, field_b='c')*",  # noqa
+                "*field_b=SimpleDataObject(field_a=2, field_b='c')) != ComplexDataObject2(field_a=ComplexDataObject*",
```
Comment (bluetech) [inline testing/test_assertion.py:816]: Any reason why you truncated the line here? The test should preferably include the entire line.

File: testing/test_assertion.py [hunk 11]
```diff
@@ -910,7 +909,9 @@ class SimpleDataObject:
 
         lines = callequal(left, right)
         assert lines is not None
-        assert "field_d: 'a' != 'b'" in lines
+        assert (
+            "    field_d: 'a' != 'b'" in lines
+        )  # indentation in output because of nested object structure
```
Comment (bluetech) [inline testing/test_assertion.py:914]:
```suggestion
        # Indented because of nested object structure.
        assert "    field_d: 'a' != 'b'" in lines
```

Comment (Vlad-Radz) [reply testing/test_assertion.py:914]: I guess, the formatter proposed to put it into the brackets. But I will check

File: testing/test_assertion.py (hunk 10 above)
Comment (Vlad-Radz) [reply testing/test_assertion.py:816]: The reason was, that when I did a multiline string (to not keep it in one line to satisfy requirements of the formatter), it included the whitespace characters from the beginning of the line. E.g.:
When I did from:
`"*field_b: SimpleDataObject(field_a=2, field_b='c') != SimpleDataObject(field_a=3, field_b='c')*"`
the following:
`"*field_b: SimpleDataObject(field_a=2, field_b='c') != \
SimpleDataObject(field_a=3, field_b='c')*"`
The output at the ends was something like:
`"*field_b: SimpleDataObject(field_a=2, field_b='c') !=                                                    SimpleDataObject(field_a=3, field_b='c')*"`,
which made the mapping not possible.

Comment (nicoddemus) [reply testing/test_assertion.py:816]: Hmmm we should probably avoid splitting the discussion over two PRs (one closed even at that).

Instead of continuing to reply here, how about copying the original message as a "quote" and link to the comment as a new post in the new PR?

Comment (bluetech) [reply testing/test_assertion.py:816]: > to not keep it in one line to satisfy requirements of the formatter

So the preferable way to handle it in this case is to just tell the linter/formatter "I know what I'm doing".

For flake8, you do this with `# noqa` comments.

For black (the formatter), you do this by delineating the code black as not auto-formatted:

```py
# fmt: off
... Your code here -- will not be auto formatted ...
# fmt: on
```

Comment (bluetech) [reply testing/test_assertion.py:816]: > Instead of continuing to reply here, how about copying the original message as a "quote" and link to the comment as a new post in the new PR?

TBH I am too lazy to do that :sleeping:

Comment (bluetech) [reply testing/test_assertion.py:816]: > to not keep it in one line to satisfy requirements of the formatter

So the preferable way to handle it in this case is to just tell the linter/formatter "I know what I'm doing".

For flake8, you do this with `# noqa` comments.

For black (the formatter), you do this by delineating the code black as not auto-formatted:

```py
# fmt: off
... Your code here -- will not be auto formatted ...
# fmt: on
```
